src/libraries/System.Linq/src/System/Linq/ElementAt.cs

```diff
@@ -63,33 +80,160 @@ public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, int i
+                {
+                    return list[index];
+                }
+            }
+            else if (TryGetElement(source, index, out TSource? element))
+            {
+                return element;
+            }
+
+            return default;
+        }
+
+        /// <summary>Returns the element at a specified index in a sequence or a default value if the index is out of range.</summary>
+        /// <param name="source">An <see cref="IEnumerable{T}" /> to return an element from.</param>
+        /// <param name="index">The index of the element to retrieve, which is either from the start or the end.</param>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="source" /> is <see langword="null" />.
+        /// </exception>
+        /// <returns>
+        ///   <see langword="default" /> if <paramref name="index" /> is outside the bounds of the <paramref name="source" /> sequence; otherwise, the element at the specified position in the <paramref name="source" /> sequence.
+        /// </returns>
+        public static TSource? ElementAtOrDefault<TSource>(this IEnumerable<TSource> source, Index index)
+        {
+            if (source == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+
+            if (!index.IsFromEnd)
+            {
+                return source.ElementAtOrDefault(index.Value);
+            }
+
+            TryGetElementFromEnd(source, index.Value, out TSource? element);
+            return element;
+        }
+
+        private static bool TryGetElement<TSource>(IEnumerable<TSource> source, int index, [MaybeNullWhen(false)] out TSource element)
+        {
+            Debug.Assert(source != null);
+
             if (index >= 0)
             {
-                if (source is IList<TSource> list)
+                using IEnumerator<TSource> e = source.GetEnumerator();
+                while (e.MoveNext())
                 {
-                    if (index < list.Count)
+                    if (index == 0)
                     {
-                        return list[index];
+                        element = e.Current;
+                        return true;
                     }
+
+                    index--;
                 }
-                else
+            }
+
+            element = default;
+            return false;
+        }
+
+        private static bool TryGetElementFromEnd<TSource>(IEnumerable<TSource> source, int indexFromEnd, [MaybeNullWhen(false)] out TSource element)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(indexFromEnd >= 0);
+
+            element = default;
+            if (indexFromEnd == 0)
+            {
+                return false;
+            }
+
+            if (source is IPartition<TSource> partition)
```

Comment:
Heads up, I just merged #48239 so it might be possible to simplify this by writing:
```csharp
if (source.TryGetNonEnumeratedCount(out int count))
{
     return TryGetElement(source, count - indexFromEnd, out element);
}
```

Reply:
Updated the code to consolidate everything into one `TryGetElementAt` method. I am not using `TryGetNonEnumeratedCount`, because using it we have to detect some types twice.

Reply:
Strictly speaking, the two methods test for different types; `TryGetNonEnumeratedCount` tests for `ICollection`/`ICollection<T>` and `IListProvider<T>` whereas `TryGetElementAt` checks for `IList<T>` and `IPartition<T>`. From the perspective of "from end" lookups all of the above are essential, however the former are not needed for regular indices. I would recommend making the following changes:

1. Keep separate private methods for regular indices and fromEnd indices. Regular indices need not check if `fromEnd` is true or test for `ICollection`/`ICollection<T>`/`IListProvider<T>` types.
2. Use `Enumerable.TryGetNonEnumeratedCount` to deduce the count of the enumerable.

Perhaps it's easier to demonstrate using actual code, so I've created a [branch](https://github.com/eiriktsarpalis/runtime/commit/f25c88d90e1bf1580e4b42143453dcf7321c78fe) to demonstrate what I mean. Disclaimer, code knocked together for illustrative purposes and has not been tested at all.

Reply:
Updated the code in a similar way, and kept the original behavior.